sled-agent/src/common/instance.rs

```diff
@@ -97,266 +131,345 @@ impl ObservedPropolisState {
+    /// position and updates the generation number, but does not clear the
+    /// destination ID or the active migration ID. This promotes a migration
+    /// target VMM into the active position without actually allowing a new
+    /// migration to begin.
+    ///
+    /// This routine should only be called when
+    /// `instance.dst_propolis_id.is_some()`.
+    fn switch_propolis_id_to_target(&mut self, now: DateTime<Utc>) {
+        assert!(self.instance.dst_propolis_id.is_some());
+
+        self.instance.propolis_id = self.instance.dst_propolis_id;
+        self.instance.gen = self.instance.gen.next();
+        self.instance.time_updated = now;
+    }
+
+    /// Forcibly transitions this instance's VMM into the specified `next`
+    /// state and updates its generation number.
+    pub(crate) fn transition_vmm(
+        &mut self,
+        next: PublishedVmmState,
+        now: DateTime<Utc>,
+    ) {
+        self.vmm.state = next.into();
+        self.vmm.gen = self.vmm.gen.next();
+        self.vmm.time_updated = now;
+    }
+
+    /// Updates the state of this instance in response to a rude termination of
+    /// its Propolis zone, marking the VMM as destroyed and applying any
+    /// consequent state updates.
+    ///
+    /// # Synchronization
+    ///
+    /// A caller who is rudely terminating a Propolis zone must hold locks
+    /// sufficient to ensure that no other Propolis observations arrive in the
+    /// transaction that terminates the zone and then calls this function.
+    ///
+    /// This function does not synchronize with ongoing live migrations. Sled
+    /// agent's callers are responsible for tearing down all of a migrating
+    /// instance's VMMs in a safe order when asking to rudely terminate a VMM
+    /// (targets before sources).
+    pub(crate) fn terminate_rudely(&mut self) {
+        // Construct a fake Propolis observation that reports that the VMM is
+        // destroyed.
+        //
+        // Doing this with a live migration in progress is perilous because of
+        // the following race:
+        //
+        // 1. The source and target VMMs agree the migration has completed
+        //    successfully.
+        // 2. Before either sled agent observes that the migration has finished,
+        //    the source is rudely terminated.
+        // 3. The instance is now running in the target VMM, but the source
+        //    sled, thinking that it still hosts the active VMM, clears the
+        //    active Propolis ID and migration IDs.
+        //
+        // There is no way to resolve this problem in the sled agent; callers
+        // who want to terminate an instance's VMMs need to do it from the
+        // "outside in" and terminate migration targets before migration
+        // sources to avoid this issue (or terminate VMMs only in situations
+        // where it is known that they cannot be participating in a migration).
```

Comment:
One other idea for consideration, if we consider this whole aspect of "sled agent destroyed dead propolis" a risk:

We could punt it to Nexus? Specifically, rather than than Sled Agents autonomously destroying VMMs / their zones, they could set a flag, signal to Nexus that something bad has happened, and let Nexus be responsible for sending a follow-up message to destroy the zone. It's a bit more of a round-trip, but it seems like (1) it would give Nexus a chance to avoid some of the race conditions you've observed, and (2) it would give us a better-defined period of time to extract whatever info we want from the zone before deleting it.

All that said, this is an idea, I don't think it should block the PR. But I noticed the set of "autonomous sled agent actions performed in response to propolis" has whittled down to only "Destroy", so removing that last one might make data flow a bit easier.

Reply:
I think I had a somewhat different case in mind when I wrote this comment: suppose we finally have a "force-quit this instance" API, and you call it on a migrating instance, and it starts by terminating the current active Propolis (i.e. the migration source). Then you could hit the race described here--the instance generation numbers are

Thinking about that some more, though, I think I would solve it differently, essentially by having some kind of rule that says something like "instance updates from sled agent are only applicable if the instance still points to the associated VMM ID." Then Nexus can "disown" all the running VMMs in a single transaction (by nulling out the UUIDs and writing an instance state that forbids the instance from restarting) and then make sure the individual VMMs have stopped. (You'd probably have to do this in a saga.)

Another approach I've considered (but not worked out in detail) is related to the "who owns the fallback state?" question you raised in other comments: instead of having sled agent send modified Instance records to Nexus, have it send more abstract messages about what a VMM has done ("VMM says it migrated out," "VMM went to the Great Control Plane in the Sky, etc.") and let Nexus own all the logic that touches Instance records. This would neatly answer the question about who owns the Instance state field (it's in the Instance so Nexus does!) but we then have to reason about, say, parallel messages about the same instance arriving from different sleds at the same time. Aren't distributed systems fun?

---

With respect to instances that actually fail (instead of being force-stopped from the API): I want to more or less completely revisit how we use the Failed state today and how we distinguish cases like "the sled reported this VMM had a problem" from "uh-oh, I couldn't talk to the sled, maybe there is a problem?" I think what you're describing here might fall into that. This ties into the comments in `handle_instance_put_result` for which I need to file an issue to get a number to put into the TODOs.